src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_pix2pix_zero.py

```diff
@@ -0,0 +1,814 @@
+                    for name, module in self.unet.named_modules():
+                        module_name = type(module).__name__
+                        if module_name == "CrossAttention" and "attn2" in name:
+                            attn_mask = module.attn_probs  # size is (num_channels, s*s, max_prompt_length)
+                            ref_xa_maps[t.item()][name] = attn_mask.detach().cpu()
+
+                    # perform guidance
+                    if do_classifier_free_guidance:
+                        noise_pred_uncond, noise_pred_text = noise_pred.chunk(2)
+                        noise_pred = noise_pred_uncond + guidance_scale * (noise_pred_text - noise_pred_uncond)
+
+                    # compute the previous noisy sample x_t -> x_t-1
+                    latents = self.scheduler.step(noise_pred, t, latents, **extra_step_kwargs).prev_sample
+
+                    # call the callback, if provided
+                    if i == len(timesteps) - 1 or ((i + 1) > num_warmup_steps and (i + 1) % self.scheduler.order == 0):
+                        progress_bar.update()
+                        if callback is not None and i % callback_steps == 0:
+                            callback(i, t, latents)
+
+        # 8. Compute the edit directions.
+        edit_direction = construct_direction(source_embedding_path, target_embedding_path)
+
+        # 9. Edit the prompt embeddings as per the edit directions discovered.
+        prompt_embeds_edit = prompt_embeds.clone()
+        prompt_embeds_edit[1:2] += edit_direction
+
+        # 10. Second denoising loop to generate the edited image.
+        latents = latents_init
+        num_warmup_steps = len(timesteps) - num_inference_steps * self.scheduler.order
+        with self.progress_bar(total=num_inference_steps) as progress_bar:
+            for i, t in enumerate(timesteps):
+                # expand the latents if we are doing classifier free guidance
+                latent_model_input = torch.cat([latents] * 2) if do_classifier_free_guidance else latents
+                latent_model_input = self.scheduler.scale_model_input(latent_model_input, t)
+
+                x_in = latent_model_input.detach().clone()
+                x_in.requires_grad = True
+
+                opt = torch.optim.SGD([x_in], lr=cross_attention_guidance_amount)
+
+                # predict the noise residual
+                noise_pred = self.unet(
+                    x_in,
+                    t,
+                    encoder_hidden_states=prompt_embeds_edit.detach(),
+                    cross_attention_kwargs=cross_attention_kwargs,
+                ).sample
+
+                loss = 0.0
+                for name, module in self.unet.named_modules():
+                    module_name = type(module).__name__
+                    if module_name == "CrossAttention" and "attn2" in name:
+                        curr = module.attn_probs
+                        ref = ref_xa_maps[t.item()][name].detach().to(device)
```

Comment:
I think this computation could / should happen directly in a "Pix2Pix-Zero"-specific attention processor. E.g. you can write a processor such as:

```python
Pix2PixZeroProcessor:
    
    def __init__(self):
        self.xa_map = {}

    def forward(...., loss=None, timestep=None):
        if loss in None:
             self.xa_map[timestep] = attn_probs
        else:
             prev_attn_probs = self.xa_map.pop(timestep)
             loss += # compute loss
```

This processor would replace the `CrossAttnProcessor` by:
- Accepting both a loss tensor and a timestep tensor that can be passed to the unet via `cross_attn_kwargs={}`
- The `Pix2PixZeroProcessor` would only be used for attention procs that have `attn2` in their name. You could do this quite easily with the https://huggingface.co/docs/diffusers/main/en/api/models#diffusers.UNet2DConditionModel.set_attn_processor function (you can pass a dict here of modue keys where you can define the class you set for every key. E.g. every `attn1` key would just stay a `CrossAttentionProc` processor, every `attn2` would become a `Pix2PixZeroProcessor` class. 

This would help to remove all this hacky code and you should be able to just do:

```python
loss = torch.tensor(0.0)

for t in timesteps:
    unet(x_in, t, encoder_hidden_states=...cross_attention_kwargs={"timesteps": t, "loss": loss})
```

Let me know if this doesn't make any sense :sweat_smile: - happy to explain in more detail, but I think here is a good example of using a more customized attn processor

Reply:
Thanks for sharing this! 

For the second point I guess you meant something like this:

```py
pix2pix_zero_attn_procs = {}

for name in unet.attn_processors.keys():
    if "attn2" in name:
        pix2pix_zero_attn_procs[name] = Pix2PixZeroCrossAttnProcessor()
    else:
        pix2pix_zero_attn_procs[name] = CrossAttnProcessor()

unet.set_attn_processor(pix2pix_zero_attn_procs)
```

(posting the snippet for future contributors)

Reply:
If we follow this design, then it would lead to:

```
TypeError: __call__() got an unexpected keyword argument 'timestep'
```

Here, we are using `Pix2PixZeroCrossAttnProcessor` for processing modules containing `attn2`, and the rest of the attention-related modules are still being processed by the default `CrossAttnProcessor`, which does not accept the `timestep` argument. This is why the error. 

Since we need to obtain the reference attention maps for Pix2Pix Zero, we need to rejig the UNet (setting attention processors, etc.) before entering into any of the denoising loops. 

I have a rough implementation [here](https://colab.research.google.com/gist/sayakpaul/746b2a08a35aa68e7cc57d4f23c523d8/scratchpad.ipynb) to test this design in isolation. Let me know if anything is unclear.

Reply:
Alright, the latest changes should be good to go (apart from some minor design-related nits) to aid this design. LMWYT.